### PR TITLE
PR 3 — Null-guarded negated operators and `is_null`/`not_null`

### PR DESCRIPTION
**Pull request by rcreasi**
Fixes a latent bug where entities with missing data were incorrectly matched by negated filter operators, and adds explicit null-checking operators since the old implicit null-matching path is now closed.

Builds on PR 2.

## Changes

- **`depmap_compute_embed/context.py`**:
  - New `is_null` / `not_null` unary operators.
  - New `_resolve_complements` preprocessing pass, called from `__init__` after slice loading. Desugars `!in`, `!has_any`, `!=`, and the `complement` node into null-guarded negations (`and` of `not_null` guards plus `!` of the positive form). Applies per variable reference in the expression.
  - Supporting helpers: `_collect_vars`, `_validate_var_refs`, and the `_NEGATED_OPS` table.
- **`tests/depmap_compute_embed/test_context.py`**: adds 15 tests covering the null behavior of each operator, the `complement` node, `is_null`/`not_null`, and nested expressions.

## Behavior change (release notes)

Previously, filters using "is not", "is not in list", or "has none of" could include items with no data for the selected property. For example, filtering for cell lines where lineage "is not" Breast would include cell lines with no lineage information at all, even though we don't actually know their lineage.

Items with missing data are now excluded from these filters. When using these operators to define groups (e.g., in a two-class comparison), items with missing data will fall into the "Other" category rather than being assigned to either group. Two new operators, "has no value" and "has a value", are available for filters that need to explicitly find items with missing data.

## How to test

```
pytest breadbox/tests/depmap_compute_embed/test_context.py
```

### CODE REVIEW

File: breadbox/breadbox/depmap_compute_embed/context.py [hunk 1]
```diff
@@ -272,3 +290,43 @@ def _validate_var_refs(expr, slice_data: dict):
...
+
+# Negated operators that should be desugared into null-guarded negations.
+# This prevents null values from being incorrectly matched by negated operators
+# (e.g. null !in ["Breast", "Lung"] would otherwise evaluate to True).
+_NEGATED_OPS = {"!in": "in", "!has_any": "has_any", "!=": "=="}
+
+
+def _resolve_complements(expr):
+    """
+    Desugars negated operators (!in, !has_any, !=) and explicit "complement"
```
Comment (pgm) [inline breadbox/breadbox/depmap_compute_embed/context.py:337]: The use of "desugars" threw me off at first (I was expecting it was doing some term rewriting between equivalent forms) and I had to think about what was going on in here. It's actually "fixing" the negated expressions to work properly with nulls. I think I'd describe this as something more like, "Added null-guards to all negations"

I have to admit this is very different way to solve the problem then I would have come up with. But I think this solves the problem and is less invasive then what I probably would have come up with.

Comment (pgm) [reply breadbox/breadbox/depmap_compute_embed/context.py:337]: Ah, I found the desugaring! 

```
if op in _NEGATED_OPS:
            positive_op = _NEGATED_OPS[op]
            ...
            negated = {"!": {positive_op: inner}}
```

Okay, I think I get it now.

File: breadbox/breadbox/depmap_compute_embed/context.py [hunk 2]
```diff
@@ -272,3 +290,43 @@ def _validate_var_refs(expr, slice_data: dict):
...
+            positive_op = _NEGATED_OPS[op]
+            inner = _resolve_complements(expr[op])
+            vars_used = sorted(_collect_vars(inner) - {"given_id"})
+            null_guards = [{"not_null": [{"var": v}]} for v in vars_used]
+            negated = {"!": {positive_op: inner}}
+            if not null_guards:
+                return negated
+            return {"and": [*null_guards, negated]}
+        if op == "complement":
+            inner = _resolve_complements(expr["complement"])
```
Comment (pgm) [inline breadbox/breadbox/depmap_compute_embed/context.py:403]: Does this correctly handle something like `{"complement": {"null: [{"var": "x"}]}}`?

If I read this correctly, that would become
```
{"complement": {"and: [{"not_null": [{"var": "x"}]}, {null: [{"var": "x"}]}}]
```
(or something like that) Makes my head hurt trying to map it out -- but my point is, `is_null` is the only operator which doesn't return false when evaluated on null. Doesn't this need to watch out for that edge case? ie: You wouldn't want a null guard on a is_null, would you?

**Review by pgm — APPROVED** [review]
One edge case I flagged. I'm not entirely sure it's a probably. Probably making a small test case for that one scenario would confirm there's a bug or prove that case is properly handled.

File: breadbox/breadbox/depmap_compute_embed/context.py (hunk 1 above)
Comment (rcreasi) [reply breadbox/breadbox/depmap_compute_embed/context.py:337]: Maybe it used the word "desugars" because `!in` and `!has_any` are operators I invented and are merely sugar for `{ "!": { "in": ... } }` and `{ "!": { "has_any": ... } }`, respectively. But it is certainly weird to call `!=` "sugar."

File: breadbox/breadbox/depmap_compute_embed/context.py (hunk 2 above)
Comment (rcreasi) [reply breadbox/breadbox/depmap_compute_embed/context.py:403]: Right, you wouldn't want a null guard on `is_null` because the only point of that operator is to give the user an "escape hatch" where they can assert nullness explicitly.

File: breadbox/breadbox/depmap_compute_embed/context.py (hunk 1 above)
Comment (rcreasi) [reply breadbox/breadbox/depmap_compute_embed/context.py:337]: And yes, Claude came up with all sorts of convoluted solutions until I told it to add a `"complement"` operator whose sole job is to compute all the null-safe negations.

File: breadbox/breadbox/depmap_compute_embed/context.py (hunk 2 above)
Comment (rcreasi) [reply breadbox/breadbox/depmap_compute_embed/context.py:403]: I'll add a test to make sure that case is handled.

Comment (rcreasi) [reply breadbox/breadbox/depmap_compute_embed/context.py:403]: @pgm Plaudits from Claude

claude-1

claude-2

Comment (pgm) [reply breadbox/breadbox/depmap_compute_embed/context.py:403]: Sometimes I feel good about myself when Claude confirms my ideas. (And then I remember it has a tendency to always say the user is right. ;) )

Regardless, it does make me feel like I'm contributing something which is nice. All these years of programming aren't totally obsolete (yet).
